packages/runtime/container-runtime/src/pendingStateManager.ts

```diff
@@ -165,6 +192,34 @@ export class PendingStateManager {
         this.pendingStates.push(pendingFlush);
     }
 
+    public processRemoteMessage(message: ISequencedDocumentMessage) {
+        while (!this.initialStates.isEmpty()) {
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            const nextMessage = this.initialStates.peekFront()!;
+            assert(nextMessage.type === "message"); // assume it's not flush for now
+            if (nextMessage.referenceSequenceNumber > message.sequenceNumber) {
+                break;
+            }
+            const context = (this.containerRuntime as any).dataStores.contexts.get(nextMessage.content?.address);
```

Comment:
This does not look right. We should not assume any structure here, i.e. that particular implementation of data store is used. It can be anything as long as it implements proper interface, and may not have contexts on it.
The right approach here is to let data store deal with it. In future, we will convert all data stores and DDSs into channels (i.e. same interface)), and possibly maybe even will allow DDS to be a child of container runtime directly, as well as having deeper nested data stores. So we should treat them as interfaces that can pass data deeper...

Reply:
I guess in practical terms - data store interface (IFluidDataStoreChannel?) should have rebaseOp method.